File: 0076. Minimum Window Substring.java
// Hard 
// Sliding Window
// Time:O(n),Space:O(k)
// https://leetcode.cn/problems/minimum-window-substring/

import java.util.*;
class Solution {
    public String minWindow(String s, String t) {
        // basecase
        if (s == null || t == null || s.length() == 0 || t.length() == 0 || s.length() < t.length())
            return "";

        Map<Character, Integer> tMap = new HashMap<>();
        for (char c : t.toCharArray()) {
            tMap.put(c, tMap.getOrDefault(c, 0) + 1);
        }

        int left = 0, minLeft = 0, minLen = Integer.MAX_VALUE, cnt = 0;
        Map<Character, Integer> window = new HashMap<>();

        for (int right = 0; right < s.length(); right++) {
            char rightChar = s.charAt(right);
            window.put(rightChar, window.getOrDefault(rightChar, 0) + 1);

            if (tMap.containsKey(rightChar) && window.get(rightChar) <= tMap.get(rightChar)) {
                cnt++;
            }

            while (cnt == t.length()) {
                if (right - left + 1 < minLen) {
                    minLen = right - left + 1;
                    minLeft = left;
                }

                char leftChar = s.charAt(left);
                window.put(leftChar, window.get(leftChar) - 1);
                if (tMap.containsKey(leftChar) && window.get(leftChar) < tMap.get(leftChar)) {
                    cnt--;
                }
                left++;
            }
        }
        return minLen == Integer.MAX_VALUE ? "" : s.substring(minLeft, minLeft + minLen);
    }
}

File: 0001-0500/0076. Minimum Window Substring.py
# Time:O(n),Space:O(k)

from collections import Counter

class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if len(s) < len(t):
            return ""
        
        need = Counter(t)

        need_count = len(need)
        
        left = 0
        min_len = float('inf')
        min_start = 0

        formed = 0

        window_count = {}

        for right in range(len(s)):
            char = s[right]
            
            window_count[char] = window_count.get(char, 0) + 1

            if char in need and window_count[char] == need[char]:
                formed += 1

            while formed == need_count and left <= right:
                window_len = right - left + 1
                if window_len < min_len:
                    min_len = window_len
                    min_start = left
                
                left_char = s[left]
                window_count[left_char] -= 1

                if left_char in need and window_count[left_char] < need[left_char]:
                    formed -= 1
                left += 1
        
        if min_len == float('inf'):
            return ""
        
        return s[min_start:min_start + min_len]
        